Remove hacky_wrapper from BackendSelect kernels

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44158 VariableKernel calls into scattered C++ api
* #44087 Add faithful C++ API
* #44071 Trace scattered tensor options arguments
* **#44062 Remove hacky_wrapper from BackendSelect kernels**
* #44005 Remove hacky_wrapper from VariableType and TraceType

Previously, BackendSelect kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable. This caused a re-wrapping step. Calling into a BackencSelect kernel required taking the individual scattered arguments, packing them into a TensorOptions, and the kernel itself then gathered them again for redispatch.

Now with this PR, BackendSelect kernels are written in the new way and no hacky_wrapper or rewrapping is needed for them.

Differential Revision: [D23484192](https://our.internmc.facebook.com/intern/diff/D23484192/)

smessmer committed:
Remove hacky_wrapper from BackendSelect kernels

Previously, BackendSelect kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable. This caused a re-wrapping step. Calling into a BackencSelect kernel required taking the individual scattered arguments, packing them into a TensorOptions, and the kernel itself then gathered them again for redispatch.

Now with this PR, BackendSelect kernels are written in the new way and no hacky_wrapper or rewrapping is needed for them.

Differential Revision: [D23484192](https://our.internmc.facebook.com/intern/diff/D23484192/)

[ghstack-poisoned]

ezyang commented on caffe2/core/tensor.h:
```diff
@@ -70,7 +70,7 @@ class CAFFE2_API Tensor final {
       : impl_(c10::make_intrusive<TensorImpl, UndefinedTensorImpl>(
             Storage::create_legacy(device),
-            c10::computeDispatchKey(at::device(device).layout(at::kStrided)),
+            c10::computeDispatchKey(typeMetaToScalarType(get_default_dtype()), at::kStrided, device),
```
The newly added use of `get_default_dtype()` here looks very questionable. What's going on here?

ezyang replied:
Also this is caffe2::Tensor, I am supremely doubtful we should be respecting default dtype on this constructor!

bhosmer replied:
Ah - well, it's definitely easier to fix that now :P

smessmer replied:
changed it to pass in `c10::nullopt` and leave it up to `computeDispatchKey` to figure out the default. Don't want this PR to change any behaviors, not fixing the dtype handling for now.

ezyang commented on tools/codegen/gen.py:
```diff
@@ -483,19 +494,19 @@ def go(f: NativeFunction) -> Optional[str]:
             return f"""\
 // aten::{f.func}
-{legacy_dispatcher_returns_type} {name}({', '.join(a.str_with_default() for a in legacy_dispatcher_args)}) {{
+{returns_type} {name}({', '.join(str(a) for a in args)}) {{
```
Yes, OK to to drop default here, that was just for compat

ezyang commented on tools/codegen/gen.py:
```diff
@@ -474,7 +474,18 @@ def go(f: NativeFunction) -> Optional[str]:
+        if local.use_c10_dispatcher() is UseC10Dispatcher.full:
+            returns_type = dispatcher_returns_type
+            args = dispatcher_args
+            exprs = dispatcher.legacydispatcherarguments_exprs(args)
```
This is a type violation and mypy agrees.

```
AssertionError: mypy failed: tools/codegen/gen.py:482:64: error: Argument 1 to "legacydispatcherarguments_exprs" has incompatible type "Sequence[DispatcherArgument]"; expected "Sequence[LegacyDispatcherArgument]"
tools/codegen/gen.py:486:20: error: Incompatible types in assignment (expression has type "Sequence[LegacyDispatcherArgument]", variable has type "Sequence[DispatcherArgument]")
Found 2 errors in 1 file (checked 1 source file)
```

As the name of this function suggests, this should take legacy dispatcher arguments, and maps them into dispatcher expressions.

In this case, you have dispatcher arguments and you want to map them to dispatcher expressions. This should be quite trivial to do, so just make a new function in dispatcher.py for it.

smessmer replied:
fixed

ezyang commented on tools/codegen/gen.py:
```diff
@@ -474,7 +474,18 @@ def go(f: NativeFunction) -> Optional[str]:
-        dispatcher_exprs = dispatcher.legacydispatcherarguments_exprs(legacy_dispatcher_args)
+
+
+        if local.use_c10_dispatcher() is UseC10Dispatcher.full:
```
I'm wondering if you can just get rid of the `local.use_c10_dispatcher() is UseC10Dispatcher.full` casing here entirely.  Right now, in BackendSelect you are using this to control whether or not you apply `hacky_wrapper_for_legacy_signatures`. But you are 100% in charge of *generating* the BackendSelect functions in the first place. So if you just generate the BackendSelect function to have the non-legacy signature in the first place, then you don't need any of this business at all.

smessmer replied:
That's basically what I'm doing with this PR, I'm generating them with the non-legacy signature directly instead of wrapping it. But I can only do that for c10-full ops since the concept of a "non-legacy signature" only exists for c10-full ops. Non-c10-full ops always use the legacy signature in their unboxed calls. That's why I still need the casing here.

ezyang commented:
To run typecheck locally, `mypy --config mypy-strict.ini`

ezyang reviewed (changes requested):
type error

bhosmer reviewed (commented):
One nontrivial thing noted inline (the redundant defaulting logic) plus a couple other ideas.

smessmer committed:
Update on "Remove hacky_wrapper from BackendSelect kernels"


Previously, BackendSelect kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable. This caused a re-wrapping step. Calling into a BackencSelect kernel required taking the individual scattered arguments, packing them into a TensorOptions, and the kernel itself then gathered them again for redispatch.

Now with this PR, BackendSelect kernels are written in the new way and no hacky_wrapper or rewrapping is needed for them.

Differential Revision: [D23484192](https://our.internmc.facebook.com/intern/diff/D23484192/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove hacky_wrapper from BackendSelect kernels"


Previously, BackendSelect kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable. This caused a re-wrapping step. Calling into a BackencSelect kernel required taking the individual scattered arguments, packing them into a TensorOptions, and the kernel itself then gathered them again for redispatch.

Now with this PR, BackendSelect kernels are written in the new way and no hacky_wrapper or rewrapping is needed for them.

Differential Revision: [D23484192](https://our.internmc.facebook.com/intern/diff/D23484192/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove hacky_wrapper from BackendSelect kernels"


Previously, BackendSelect kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable. This caused a re-wrapping step. Calling into a BackencSelect kernel required taking the individual scattered arguments, packing them into a TensorOptions, and the kernel itself then gathered them again for redispatch.

Now with this PR, BackendSelect kernels are written in the new way and no hacky_wrapper or rewrapping is needed for them.

Differential Revision: [D23484192](https://our.internmc.facebook.com/intern/diff/D23484192/)

[ghstack-poisoned]

ezyang commented on tools/codegen/api/dispatcher.py:
```diff
@@ -107,3 +107,6 @@ def cpparguments_exprs(args: Sequence[CppArgument]) -> Sequence[DispatcherExpr]:
 def legacydispatcherarguments_exprs(args: Sequence[LegacyDispatcherArgument]) -> Sequence[DispatcherExpr]:
     return cpparguments_exprs([CppArgument(type=a.type, name=a.name, default=None, argument=a.argument) for a in args])
+
+def dispatcherarguments_exprs(args: Sequence[DispatcherArgument]) -> Sequence[DispatcherExpr]:
```
Or just call this `exprs`, I suppose :)

ezyang replied:
Don't feel strongly about this.

smessmer committed:
Update on "Remove hacky_wrapper from BackendSelect kernels"


Previously, BackendSelect kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable. This caused a re-wrapping step. Calling into a BackencSelect kernel required taking the individual scattered arguments, packing them into a TensorOptions, and the kernel itself then gathered them again for redispatch.

Now with this PR, BackendSelect kernels are written in the new way and no hacky_wrapper or rewrapping is needed for them.

Differential Revision: [D23484192](https://our.internmc.facebook.com/intern/diff/D23484192/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove hacky_wrapper from BackendSelect kernels"


Previously, BackendSelect kernels were still written in the legacy way, i.e. they took one TensorOptions argument instead of scattered dtype, layout, device, pin_memory,  and they used hacky_wrapper to be callable. This caused a re-wrapping step. Calling into a BackencSelect kernel required taking the individual scattered arguments, packing them into a TensorOptions, and the kernel itself then gathered them again for redispatch.

Now with this PR, BackendSelect kernels are written in the new way and no hacky_wrapper or rewrapping is needed for them.

Differential Revision: [D23484192](https://our.internmc.facebook.com/intern/diff/D23484192/)

[ghstack-poisoned]
